server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -278,44 +329,94 @@ private static HitContext prepareNonNestedHitContext(
         if (id == null) {
-            SearchHit hit = new SearchHit(docId);
+            SearchHit hit = new SearchHit(docId, null, null, refCountedHit);
             // TODO: can we use real pooled buffers here as well?
-            Source source = Source.lazy(lazyStoredSourceLoader(profiler, subReaderContext, subDocId));
+            Source source = Source.lazy(
+                lazyStoredSourceLoader(
+                    profiler,
+                    subReaderContext,
+                    subDocId,
+                    memAccountingRefCounted,
+                    accountMemoryWithCircuitBreaker,
+                    memoryUsageBytesAccumulator,
+                    accumulatedBytesInLeaf
+                )
+            );
             return new HitContext(hit, subReaderContext, subDocId, Map.of(), source, rankDoc);
         } else {
-            SearchHit hit = new SearchHit(docId, id);
+            SearchHit hit = new SearchHit(docId, id, null, refCountedHit);
             Source source;
             if (requiresSource) {
                 Timer timer = profiler.startLoadingSource();
                 try {
                     source = sourceLoader.source(leafStoredFieldLoader, subDocId);
+                    memoryUsageBytesAccumulator.accept(source.internalSourceRef().length());
+                    if (accountMemoryWithCircuitBreaker) {
+                        memAccountingRefCounted.account(accumulatedBytesInLeaf.get(), "fetch phase source loader");
+                    }
+                } catch (CircuitBreakingException e) {
+                    hit.decRef();
+                    throw e;
                 } finally {
                     if (timer != null) {
                         timer.stop();
                     }
                 }
             } else {
-                source = Source.lazy(lazyStoredSourceLoader(profiler, subReaderContext, subDocId));
+                source = Source.lazy(
+                    lazyStoredSourceLoader(
+                        profiler,
+                        subReaderContext,
+                        subDocId,
+                        memAccountingRefCounted,
+                        accountMemoryWithCircuitBreaker,
+                        memoryUsageBytesAccumulator,
+                        accumulatedBytesInLeaf
+                    )
+                );
             }
             return new HitContext(hit, subReaderContext, subDocId, leafStoredFieldLoader.storedFields(), source, rankDoc);
         }
     }
 
-    private static Supplier<Source> lazyStoredSourceLoader(Profiler profiler, LeafReaderContext ctx, int doc) {
+    private static Supplier<Source> lazyStoredSourceLoader(
+        Profiler profiler,
+        LeafReaderContext ctx,
+        int doc,
+        MemoryAccountingBytesRefCounted memAccountingRefCounted,
+        boolean submitToCB,
+        Consumer<Integer> memoryUsageAccumulator,
+        Supplier<Integer> accumulatedBytesInLeaf
```

Comment:
I guess we want these generic interfaces for testing, but maybe at least combine all the memory accounting stuff into a single interface and thus parameter? We're now adding 4 parameters to a 3 parameter method, just for the accounting :) Plus, we really don't want to box the ints here anyway :)

Reply:
>maybe at least combine all the memory accounting stuff into a single interface and thus parameter?

We're accumulating locally on every hit and reading what's accumulated at a different time. So we need two things here.
Dropped the boxing of ints. Great shout - had forgotten about the primitive supplier/consumers

Reply:
But why can't all of this be a single object? (we can create multiple instances, though I'd imagine we'd still only have one per thread?
Adding 4 parameters seems like a massive increase in complexity, that's all :)

Reply:
Oh, I actually might have a way to unify these